### ISSUE TITLE
Session scoped fixtures can leak out of a class

### ISSUE THREAD

**Issue by jinty** [pytest-dev/pytest#2145]
Best explained with a failing test, see:

https://github.com/jinty/pytest/commit/c0048b115314dd4a54d1fea5032d85f52efd221b

or inline for the lazy:

```
import pytest

@pytest.fixture(scope='session')
def foo():
    return 'foo'

@pytest.fixture(scope='session')
def bar(foo):
    return foo + 'bar'

@pytest.fixture(scope='session')
def baz(bar):
    return bar + 'baz'

class TestSpecific:

    @pytest.fixture(scope='session')
    def foo(self):
        return 'FOO'

    @pytest.fixture(scope='session')
    def baz(self, bar):
        return bar + 'BAZ'

    def test_b(self, foo, bar, baz):
        pass

def test_c(foo, bar, baz):
    assert foo == 'foo'
    assert bar == 'foobar'
    assert baz == 'foobarbaz'
```

Fails with the current master.

**Comment by lwm** [comment]
Yep, reproduced this one with `wget goo.gl/CDgb12 -O test_2145.py && pytest`.

I thought `scope=session` fixtures were global, so this is expected?

**Comment by jinty** [comment]
On Mon, Dec 19, 2016 at 04:30:29PM -0800, Luke Murphy wrote:
> Yep, reproduced this one with `wget goo.gl/CDgb12 -O test_2145.py && pytest`.
> 
> I thought `scope=session` fixtures were global, so this is expected?

At least I don't expected that the result changes depending on test
ordering and the dependency graph of the fixtures.

I would prefer an explicit error (i.e. "don't do that, it's silly")
rather than the current behaviour which took quite a while to debug.

-- 
Brian Sutherland

**Comment by RonnyPfannschmidt** [comment]
the short story is, fixture declaration and mapping declarations to locations and caches is a complete mess and needs major refactoring ^^

**Comment by nicoddemus** [comment]
While I agree with @RonnyPfannschmidt that the fixture system could use a refactoring to account for the features added in the last years, this warrants more investigation because I think there's a bug that should be solvable here; `test_c` should not have been able to see the fixtures declared by `TestSpecific`.

**Comment by LEscobar-Driver** [comment]
Is there a fix to the **ScopeMismatch** problem yet?
